compose/config/config.py

```diff
@@ -132,29 +163,42 @@ def pre_process_config(config):
     Pre validation checks and processing of the config file to interpolate env
     vars returning a config dict ready to be tested against the schema.
     """
-    config = interpolate_environment_variables(config)
-    return config
+    return interpolate_environment_variables(config)
 
 
 def load(config_details):
-    config, working_dir, filename = config_details
+    working_dir, configs = config_details
 
-    processed_config = pre_process_config(config)
-    validate_against_fields_schema(processed_config)
-
-    service_dicts = []
-
-    for service_name, service_dict in list(processed_config.items()):
-        loader = ServiceLoader(
-            working_dir=working_dir,
-            filename=filename,
-            service_name=service_name,
-            service_dict=service_dict)
+    def build_service(filename, service_name, service_dict):
+        loader = ServiceLoader(working_dir, filename, service_name, service_dict)
         service_dict = loader.make_service_dict()
         validate_paths(service_dict)
-        service_dicts.append(service_dict)
-
-    return service_dicts
+        return service_dict
+
+    def load_file(filename, config):
+        processed_config = pre_process_config(config)
+        validate_against_fields_schema(processed_config)
+        return [
+            build_service(filename, name, service_config)
+            for name, service_config in processed_config.items()
+        ]
+
+    def merge_services(base, override):
+        return {
+            name: merge_service_dicts(base.get(name, {}), override.get(name, {}))
+            for name in set(base) | set(override)
+        }
+
+    def combine_configs(base, override):
+        service_dicts = load_file(base.filename, base.config)
+        if not override:
+            return service_dicts
+
+        return ConfigFile(
+            override.filename,
+            merge_services(base.config, override.config))
+
+    return reduce(combine_configs, configs + [None])
```

Comment:
> Also, we should probably document the attributes (including types) of both ConfigDetails and ConfigFile.

We don't really do this in a consistent way right now. I'd love to use [sphinx](http://sphinx-doc.org/) for all the api docs so the documentation is consistent and machine readable.